File: scripts/set-webhook.ts
/**
 * Register or clear the Telegram webhook URL.
 *
 * Usage:
 *   npx tsx scripts/set-webhook.ts https://your-app.vercel.app/api/webhook
 *   npx tsx scripts/set-webhook.ts --clear
 */

const BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const WEBHOOK_SECRET = process.env.TELEGRAM_WEBHOOK_SECRET;

if (!BOT_TOKEN) {
  console.error("Missing TELEGRAM_BOT_TOKEN environment variable.");
  console.error("Set it in .env or export it before running this script.");
  process.exit(1);
}

const url = process.argv[2];
if (!url) {
  console.error("Usage: npx tsx scripts/set-webhook.ts <url>");
  console.error("       npx tsx scripts/set-webhook.ts --clear");
  process.exit(1);
}

const API_BASE = `https://api.telegram.org/bot${BOT_TOKEN}`;

async function main() {
  if (url === "--clear") {
    const res = await fetch(`${API_BASE}/deleteWebhook`);
    const data = await res.json();
    if (data.ok) {
      console.log("Webhook cleared.");
    } else {
      console.error("Failed to clear webhook:", data.description);
      process.exit(1);
    }
  } else {
    const params: Record<string, string> = { url };
    if (WEBHOOK_SECRET) {
      params.secret_token = WEBHOOK_SECRET;
    }

    const res = await fetch(`${API_BASE}/setWebhook`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(params),
    });
    const data = await res.json();

    if (data.ok) {
      console.log(`Webhook set to: ${url}`);
      console.log(`Secret token: ${WEBHOOK_SECRET ? "configured" : "none"}`);
    } else {
      console.error("Failed to set webhook:", data.description);
      process.exit(1);
    }
  }
}

main();

File: scripts/set_webhook.py
#!/usr/bin/env python3
"""Register a Telegram webhook URL for the bot.

Usage:
    python scripts/set_webhook.py https://your-app.vercel.app/api/webhook

To clear the webhook (revert to polling mode for local dev):
    python scripts/set_webhook.py --clear
"""
from __future__ import annotations

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "src"))

from telegram import Bot
from climb_bot.config import load_settings


async def set_webhook(url: str, secret: str | None) -> None:
    settings = load_settings()
    bot = Bot(settings.bot_token)

    async with bot:
        if url == "--clear":
            await bot.delete_webhook()
            print("Webhook cleared. Bot will not receive updates until a new webhook is set.")
        else:
            await bot.set_webhook(url=url, secret_token=secret)
            info = await bot.get_webhook_info()
            print(f"Webhook set to: {info.url}")
            print(f"Secret token: {'configured' if secret else 'none'}")


def main() -> None:
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    url = sys.argv[1]
    settings = load_settings()
    asyncio.run(set_webhook(url, settings.webhook_secret))


if __name__ == "__main__":
    main()
